### PR TITLE
Check if /dev/urandom is a character device

### LINKED ISSUE #1626: os.getRandomBytesDevURandom should check if /dev/urandom is a character device
Whilst looking at the Zig standard library to get a feel for the language, I noticed that [`os.getRandomBytesDevURandom`](https://github.com/ziglang/zig/blob/master/std/os.zig#L132) doesn't check whether `/dev/urandom` is a valid character device before reading it.

It may be worth looking at some other implementations of getting random data to see how it's done elsewhere:

- [PHP's `random_bytes()`](https://github.com/php/php-src/blob/master/ext/standard/random.c#L85)
- [libsodium's `sysrandom`](https://github.com/jedisct1/libsodium/blob/master/src/libsodium/randombytes/sysrandom/randombytes_sysrandom.c)

### PR DESCRIPTION
**Pull request by euantorano**
Fixes #1626 by checking if `/dev/urandom` is a valid character device before using it.

[libsodium](https://github.com/jedisct1/libsodium/blob/master/src/libsodium/randombytes/sysrandom/randombytes_sysrandom.c#L203) has a slightly different check in some cases where it checks `S_ISNAM() or S_ISCHR()`, but `S_ISNAM` (`XENIX named special file with two subtypes`) isn't currently defined in Zig's standard library and I'm not sure if we need it since it seems specific to a tiny subset of systems...

I'm also not sure what error to return in the case that `/dev/urandom` isn;t a character device - I just return `Unexpected` for now, but might it be better to add a more specific error?

### CODE REVIEW

**Comment by euantorano** [comment]
Looks like some utility functions are missing on macOS and FreeBSD. I'll add them later to fix this.

File: std/os.zig [hunk 1]
```diff
@@ -133,6 +133,11 @@ fn getRandomBytesDevURandom(buf: []u8) !void {
     const fd = try openC(c"/dev/urandom", O_RDONLY | O_CLOEXEC, 0);
     defer close(fd);
 
+    const st = try fstat(fd);
+    if (!system.S_ISCHR(st.mode)) {
```
Comment (andrewrk) [inline std/os.zig]: S_ISCHR should be in the bits.zig files, and std/os.zig is one of the few places where we use `usingnamespace` on all the system bits. In this particular file (but not others) the convention is to omit the namespace.
```suggestion
    if (!S_ISCHR(st.mode)) {
```

File: std/os.zig [hunk 2]
```diff
@@ -133,6 +133,11 @@ fn getRandomBytesDevURandom(buf: []u8) !void {
...
+        return error.Unexpected;
```
Comment (andrewrk) [inline std/os.zig]: Quoting the docs for the `Unexpected` error:

```zig
    /// The Operating System returned an undocumented error code.
    /// This error is in theory not possible, but it would be better
    /// to handle this error than to invoke undefined behavior.
```

Let's come up with an appropriate name for this error. We could either use one of the error codes from `OpenError` such as `FileNotFound` or `NoDevice`, or add a unique error code here if it is desirable to handle this situation differently. My suggestion follows:

```suggestion
        return error.NoDevice;
```

File: std/os.zig [hunk 3]
```diff
@@ -133,6 +133,11 @@ fn getRandomBytesDevURandom(buf: []u8) !void {
...
+    }
+
     const stream = &std.fs.File.openHandle(fd).inStream().stream;
     stream.readNoEof(buf) catch return error.Unexpected;
```
Comment (andrewrk) [inline std/os.zig]: Note that the reason this here is `Unexpected` is that it is in theory impossible for this situation to happen.

File: std/os/darwin.zig [hunk 4]
```diff
@@ -5,3 +5,4 @@ pub const is_the_target = switch (builtin.os) {
     else => false,
 };
 pub usingnamespace std.c;
+pub usingnamespace @import("bits.zig");
```
Comment (andrewrk) [inline std/os/darwin.zig]: why is this added?

**Comment by andrewrk** [comment]
> Looks like some utility functions are missing on macOS and FreeBSD. I'll add them later to fix this.

Can you confirm that you got these definitions from the respective operating systems' .h files, not just copy pasted?

**Comment by euantorano** [comment]
> Can you confirm that you got these definitions from the respective operating systems' .h files, not just copy pasted?

Yep. The constant values are also documented in publicly available man pages:

- FreeBSD (scroll down to the documentation about the `st_mode`): https://www.freebsd.org/cgi/man.cgi?query=stat&sektion=2
- macOS (again, scroll down to the documentation about the `st_mode`): https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/stat.2.html and https://developer.apple.com/library/archive/documentation/System/Conceptual/ManPages_iPhoneOS/man2/chmod.2.html#//apple_ref/doc/man/2/chmod (documentation about the list of access modes)

File: std/os/darwin.zig (hunk 4 above)
Comment (euantorano) [reply std/os/darwin.zig]: This was copied from the Linux file, as `S_ISCHR` was coming in from the `bits`: https://github.com/ziglang/zig/blob/master/std/os/linux.zig#L22 - I also added it to `os/freebsd.zig`. It can be removed from both.

Comment (andrewrk) [reply std/os/darwin.zig]: Oh, hmm. Yes I think what you have here is correct. Thanks!
